### ISSUE TITLE
[lock] MultiMap lock - Thread is getting stuck on call of lock(key)

### ISSUE THREAD

**Issue by pramodterdale** [hazelcast/hazelcast#9055]
I am using MultiMap implementation for my use case, I need to use MultiMap.lock(key) feature to make sure same key is not getting modified concurrently, i have noticed that the lock(key) method gets called and randomly the thread does not acquire a lock and get stuck forever, I have seen this behavior consistently but on random keys, no specific keys pattern. Most of the keys are unique, i would say 99% of them are unique but 1% of them may be duplicate and need to be handled in multi-threaded env and hence using lock(key) feature. My key column is of type String and using hazelcast and client with version 3.7

```
String key = getKey();
MultiMap<String, String> mmap = hz.getMultiMap("SampleMap");
mmap.lock(key)  //This line is where thread gets stuck and does not come out
//Do something
mmpa.unlock(key);
```

Is here any issue with lock feature of MultiMap, any help is very much appreciated..?

**Comment by jerrinot** [comment]
hi @pramodterdale,

is there a chance someone other thread fails to unlock the key? are you always unlocking it a `finally` block?

**Comment by pramodterdale** [comment]
Yes, in finally block i unlock it, if it fails to unlock in finally block i force unlock it using try/catch block.
 I read the documentation and it says locks are re-entrant so thread get stuck untill it acquires the lock, in this case it is trying but not sure why it is not able to acquire lock even though i force unlock it in case it is not able to unlock.

**Comment by mmedenjak** [comment]
@pramodterdale Is this something that you can reproduce easily? Do you know or can you check if some other thread can acquire the lock even when one thread gets stuck on `lock`? Meaning, can some other thread continue even though one is blocked?
 
Can you also try using the timed lock version - `MultiMap#tryLock(K, long, java.util.concurrent.TimeUnit)`? Set the timeout to some value higher than any expected for the business logic duration but short enough that you can observe the thread which failed to acquire the lock. Does the thread unblock after the timeout or does it stay blocked indefinitely?

**Comment by mmedenjak** [comment]
@pramodterdale Did you have time to try any of the suggestions?

**Comment by jerrinot** [comment]
I am moving this ticket into 3.9 as we do not have enough data to act on it as this moment.

**Comment by mmedenjak** [comment]
@pramodterdale are you still experiencing this issue? Do you know if there are any membership changes in the cluster?

**Comment by mmedenjak** [comment]
@pramodterdale I am closing this issue since we are unable to reproduce it at this moment. If you are still experiencing the issue, you can :
- try some of the suggestions
- try upgrading to hazelcast 3.8.4
If the problem persists, please open a new issue.
